File: LeetCode/439-ternary-expression-parser/ex_439.java
public class Solution {
    public String parseTernary(String expression) {
        if (expression.length() == 1)   return expression;
        int idx = findSplitIndex(expression);
        if (expression.charAt(0) == 'T')
            return parseTernary(expression.substring(2, idx));
        else
            return parseTernary(expression.substring(idx + 1));
    }
    
    private int findSplitIndex(String exp) {
        int count = -1, i = 2;
        while (i < exp.length()) {
            if (exp.charAt(i) == '?')
                count--;
            else if (exp.charAt(i) == ':') {
                count++;
                if (count == 0) break;
            }
            i++;
        }
        return i;
    }
}

File: LeetCode/439-ternary-expression-parser/ex_439.py
class Solution(object):
    def parseTernary(self, expression):
        """
        :type expression: str
        :rtype: str
        """
        if len(expression) == 1:    return expression
        idx = self.findSplitIndex(expression)
        if expression[0] == 'T':
            return self.parseTernary(expression[2:idx])
        else:
            return self.parseTernary(expression[idx+1:])
    
    def findSplitIndex(self, exp):
        count = -1
        i = 2
        while i < len(exp):
            if exp[i] == '?':
                count -= 1
            elif exp[i] == ':':
                count += 1
                if count == 0:  break
            i += 1
        return i